### PR TITLE
Illumos/Solaris updates

### LINKED ISSUE #12213: Cannot build stage 2 on x86-64-illumos
### Zig Version

0.10.0-dev.3264+d99c32c2e

### Steps to Reproduce

1. Build LLVM 14.0.6 with [this](https://www.sgregoratto.me/paste/llvm-14-illumos.patch) patch applied. It works around some rough spots that haven't been fixed in Solaris. You can use the method found in the wiki, but the script I use is:
    ```sh
    #!/bin/sh
    cmake "llvm" \
        -B build -G Ninja \
        -DLLVM_ENABLE_PROJECTS="lld;clang" \
        -DLLVM_ENABLE_LIBXML2=OFF \
        -DCMAKE_INSTALL_PREFIX="$HOME/local" \
        -DCMAKE_PREFIX_PATH="$HOME/local" \
        -DLLVM_INCLUDE_TESTS=OFF \
        -DLLVM_INCLUDE_GO_TESTS=OFF \
        -DLLVM_INCLUDE_EXAMPLES=OFF \
        -DLLVM_INCLUDE_BENCHMARKS=OFF \
        -DLLVM_BUILD_LLVM_DYLIB=ON \
        -DLLVM_ENABLE_BINDINGS=OFF \
        -DLLVM_ENABLE_OCAMLDOC=OFF \
        -DLLVM_ENABLE_Z3_SOLVER=OFF \
        -DCLANG_BUILD_TOOLS=OFF \
        -DCLANG_INCLUDE_DOCS=OFF \
        -DLLVM_INCLUDE_DOCS=OFF \
        -DCMAKE_BUILD_TYPE=Release
    ninja -C build install
    ```
2. Build stage 1 with this LLVM build:
    ```sh
    cmake -B build -G Ninja -DCMAKE_PREFIX_PATH="$HOME/local" -DCMAKE_INSTALL_PREFIX="$HOME/local" -DCMAKE_BUILD_TYPE=Release
    ninja -C build
    ```
3. Build stage 2:
    ```sh
    cd build
    ./zig build -p stage2 -Denable-llvm
    ```

### Expected Behavior

Stage 2 LLVM zig version to be built.

### Actual Behavior

A ton of linker errors. Running with `--verbose-link` for more info (first line) renders [this log](https://www.sgregoratto.me/paste/stage2.log) (too big for github).

Subtask of #7152.

### PR DESCRIPTION
**Pull request by rzezeski**
I'd like to integrate these changes I've made in service of getting newer Zig (0.11.0 and latest master) running on illumos. I've tested these changes on OmniOS where I'm able to compile & run the latest version of `ncdu` along with my own personal Zig project. I've also managed to run some of the test suite, albeit with failures that need fixing. I spoke to @andrewrk about some of these failures in the Zig IRC channel, and he agreed it would be okay to get my improvements up as a base to start fixing bugs.

Zig currently has no `illumos` OS tag, only `solaris`. While these two systems stem from the same codebase, they broke apart in 2010 when Oracle reclosed the source and illumos was formed as an open-source fork based on the latest code drop. After 13 years they have grown enough apart that it is detrimental to consider them the same platform. I would like to propose adding an illumos tag, probably similar to how it's been done in Go and Rust. This work will need to be done if #7152 is to be accepted. But for this initial support I've worked under the existing `solaris` tag.

These changes should fix #12213.

I've presented this as a series of smaller changes so it might be easier to digest. But I'm happy to squash this into one if that's preferred.

### CODE REVIEW

**Comment by The-King-of-Toasters** [comment]
I've built Zig with these patches on an OmniOS VM and it works great. Of course it requires building LLVM+Clang+LLD from source as OmniOS doesn't package LLD. I also found that I needed to compile stage3 with `-fPIC` to be able to link against libc.

File: lib/std/target.zig [hunk 1]
```diff
@@ -1575,7 +1575,7 @@ pub const Target = struct {
             .netbsd => return copy(&result, "/libexec/ld.elf_so"),
             .openbsd => return copy(&result, "/usr/libexec/ld.so"),
             .dragonfly => return copy(&result, "/libexec/ld-elf.so.2"),
-            .solaris => return copy(&result, "/lib/64/ld.so.1"),
+            .solaris => return copy(&result, "/usr/lib/amd64/ld.so.1"),
```
Comment (The-King-of-Toasters) [inline lib/std/target.zig:1578]: Would `/usr/lib/64/ld.so.1` also work here?

**Comment by andrewrk** [comment]
> I would like to propose adding an illumos tag

Please proceed. Start here:

```diff
--- a/lib/std/target.zig
+++ b/lib/std/target.zig
@@ -58,6 +58,7 @@ pub const Target = struct {
             glsl450,
             vulkan,
             plan9,
+            illumos,
             other,
 
             pub inline fn isDarwin(tag: Tag) bool {
```

From there, compilation errors should guide you along the way.

**Comment by The-King-of-Toasters** [comment]
Ryan, I've forked your branch and added support for the `.illumos` tag [here](https://github.com/gh-fork-dump/zig/tree/illumos-updates). I've confirmed that stage3 builds and runs the stdlib tests on the `x86_64-solaris` and `x86_64-illumos` tuples, which was about the safe functionality before. Can you pull these changes into your tree and confirm that they work on your end?

BTW, this will require rebuilding `zig1.wasm` to be able to parse the new tag (at least it did for me).

**Comment by rzezeski** [comment]
@The-King-of-Toasters We ended up doing the same work pretty much, so I cherry picked your commits.

Though I removed illumos from semver as it doesn't do any versioning at the system level. I can't remember if Oracle Solaris decided to start making use of the uname version again or not, but I decided to leave it alone.

**Comment by andrewrk** [comment]
Nice work!

**Pull request merged**